pkg/vault/contracts/Vault.sol

```diff
@@ -1234,39 +1247,30 @@ contract Vault is IVaultMain, VaultCommon, Proxy {
 
-            // Only updates buffer balances if buffer has a surplus of underlying tokens.
-            if (bufferUnderlyingSurplus > 0) {
-                if (kind == SwapKind.EXACT_IN) {
-                    // Since `bufferUnderlyingSurplus` was wrapped, the final amount out needs to discount the wrapped
-                    // amount that will stay in the buffer. Refresh `bufferWrappedSurplus` after external calls on the
-                    // wrapped token. For EXACT_IN, `vaultUnderlyingDeltaHint` and `amountInUnderlying` do not change
-                    // after the deposit operation, so the `bufferUnderlyingSurplus` does not need to be recalculated.
-                    bufferWrappedSurplus = wrappedToken.previewDeposit(bufferUnderlyingSurplus);
-                    amountOutWrapped = vaultWrappedDeltaHint - bufferWrappedSurplus;
-                } else {
-                    // If the buffer has a surplus of underlying tokens, it wraps the surplus + amountIn, so the final
-                    // amount in needs to be discounted for that. For EXACT_OUT, `vaultWrappedDeltaHint` and
-                    // `amountOutWrapped` do not change after the mint operation, so the `bufferWrappedSurplus` does
-                    // not need to be recalculated.
-                    bufferUnderlyingSurplus = wrappedToken.previewMint(bufferWrappedSurplus);
-                    amountInUnderlying = vaultUnderlyingDeltaHint - bufferUnderlyingSurplus;
-                }
+            // Only updates buffer balances if buffer has an imbalance of underlying tokens.
+            if (bufferUnderlyingImbalance != 0) {
+                // A wrap should remove underlying from the sender and buffer, and return wrapped tokens to the sender
+                // and buffer. However, `vaultUnderlyingDeltaHint` may be smaller than `amountInUnderlying`, meaning
+                // that the buffer had more wrapped than underlying. If that's the case, an amount of underlying must
+                // be added to the buffer, so `bufferUnderlyingImbalance` must be negative (note that the buffer
+                // balance update below already removes bufferUnderlyingImbalance from the buffer, so if this number
+                // is negative it'll actually be added to the buffer balances).
+                bufferUnderlyingImbalance = vaultUnderlyingDeltaHint.toInt256() - amountInUnderlying.toInt256();
+                // A wrap should remove underlying from the sender and buffer, and return wrapped tokens to the sender
+                // and buffer. However, `vaultWrappedDeltaHint` may be smaller than `amountOutWrapped`, meaning
+                // that the buffer had more wrapped than underlying. If that's the case, an amount of wrapped must
+                // be removed from the buffer, so `bufferWrappedImbalance` must be negative.
+                bufferWrappedImbalance = vaultWrappedDeltaHint.toInt256() - amountOutWrapped.toInt256();
```

Comment:
As stated above, this is the only real place where we're computing the wrapped imbalance (the one before `_settleWrap` can be computed without calling `getBufferWrappedImbalance`.

Then, this can be a local inside this `if`.